File: packages/service-core/src/config/secret.ts
import { existsSync, readFileSync } from 'node:fs'
import { join } from 'node:path'

/**
 * Where a secret comes from, and in which order — `contracts/secrets.json` is normative.
 *
 * A secret in an environment variable is readable in `/proc/<pid>/environ`, is inherited by
 * every child process, and turns up in `docker inspect`, in crash dumps and in CI logs. So the
 * environment may carry a *path* to a secret, and the best of the three sources carries no
 * environment entry at all.
 *
 * Nothing here is systemd-specific, and that is what makes the mechanism usable from both
 * implementations: `LoadCredential=` places the secret as a file in a per-unit directory on a
 * tmpfs and names that directory in `$CREDENTIALS_DIRECTORY`. Reading it is reading a file at a
 * path the process was told. There is no API to bind against and no platform check to write —
 * the variable is simply unset everywhere that is not a systemd service.
 */

/**
 * Every value this project treats as a secret.
 *
 * Declared rather than discovered, and short because the services are. A secret added to either
 * implementation is added to `contracts/secrets.json` first: an operator who has learned
 * `DB_PASSWORD_FILE` has learned the mechanism, and a second secret that resolved differently
 * would make that knowledge wrong. The mail relay's password and the JWT signing key are the
 * one this was written in anticipation of, and the JWT signing key is the one still to come.
 */
export const SECRET_VARIABLES = ['DB_PASSWORD', 'EMAIL_PASSWORD'] as const

export type SecretVariable = (typeof SECRET_VARIABLES)[number]

/**
 * Which of the three sources a secret came from.
 *
 * `setup` is the caller this exists for. It offers the current value as the default for its
 * question and writes the answer into `.env` — which is right for a value that was in the
 * environment and wrong for one that was not: pressing Enter on a secret that systemd keeps on
 * a tmpfs would copy it into a file in the working directory, and a mechanism that can be
 * downgraded by answering a prompt is not one.
 */
export type SecretSource = 'credential' | 'file' | 'environment'

/** One trailing line ending and nothing else — see contracts/secrets.json, resolution.rules. */
function stripOneLineEnding(content: string): string {
  if (content.endsWith('\r\n')) {
    return content.slice(0, -2)
  }
  return content.endsWith('\n') ? content.slice(0, -1) : content
}

/**
 * Where the secret @p name would come from, or undefined when no source has it.
 *
 * The same walk `resolveSecret` makes, reporting the source instead of the value — the two can
 * never disagree about which one wins, because they are the same three checks in the same order.
 */
export function secretSource(
  name: string,
  env: Record<string, string | undefined> = process.env,
): SecretSource | undefined {
  const credentials = env.CREDENTIALS_DIRECTORY
  if (credentials !== undefined && credentials !== '' && existsSync(join(credentials, name))) {
    return 'credential'
  }
  const path = env[`${name}_FILE`]
  if (path !== undefined && path !== '') {
    return 'file'
  }
  return env[name] === undefined ? undefined : 'environment'
}

/**
 * The secret @p name, from the best source that has it, or undefined when none has it.
 *
 * A source that is *named* and cannot be read throws rather than falling back: an unreadable
 * secret must not quietly become an empty one, because an empty password is how a process
 * connects as somebody else or fails much later about the wrong thing.
 */
function resolveSecret(name: string, env: Record<string, string | undefined>): string | undefined {
  const credentials = env.CREDENTIALS_DIRECTORY
  if (credentials !== undefined && credentials !== '') {
    try {
      return stripOneLineEnding(readFileSync(join(credentials, name), 'utf8'))
    } catch (error) {
      /* Absent is ordinary — a unit loads the credentials it needs and no others. Present and
         unreadable is not, and the difference is the errno. */
      if ((error as NodeJS.ErrnoException).code !== 'ENOENT') {
        throw new Error(
          `the systemd credential ${join(credentials, name)} could not be read: ${String(error)}`,
        )
      }
    }
  }

  const path = env[`${name}_FILE`]
  if (path !== undefined && path !== '') {
    try {
      return stripOneLineEnding(readFileSync(path, 'utf8'))
    } catch (error) {
      throw new Error(
        `${name}_FILE names ${path}, which could not be read — refusing to fall back to ${name}: ${String(error)}`,
      )
    }
  }

  return env[name]
}

/**
 * The environment with every secret resolved into it, ready to be parsed against a schema.
 *
 * A copy rather than a mutation of `process.env`: a secret that was deliberately kept out of the
 * environment has no business being put back into it, where the next child process would inherit
 * it. What the schema sees is not what `/proc/<pid>/environ` says, and that is the point.
 */
export function resolveSecrets(
  env: Record<string, string | undefined>,
): Record<string, string | undefined> {
  const resolved: Record<string, string | undefined> = { ...env }
  for (const name of SECRET_VARIABLES) {
    resolved[name] = resolveSecret(name, env)
  }
  return resolved
}

File: fast-servers/service-core/src/secret.c
/*
 * A secret out of the best source that has it. contracts/secrets.json is normative; this file is
 * its C reading.
 *
 * The `.env` half of this used to live here too and does not any more: service_core/env_file.h
 * reads that file, writes it for `setup`, and knows dotenv's quoting rules -- one reader for one
 * format, and the reference path reads the same file with dotenv itself.
 */
#include "service_core/secret.h"

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "service_core/log/log.h"

/*
 * Reads a whole file into @p out.
 *
 * One trailing "\n" or "\r\n" goes, and nothing else -- contracts/secrets.json,
 * resolution.rules: `echo secret > file` writes a newline that is not part of the secret, while
 * trimming further would make a password that ends in a space impossible to configure.
 *
 * A file that does not fit is SC_ERR_TOO_LONG rather than a truncated secret.
 *
 * @p absent separates the two ways a read fails, and the separation is the contract's: a file
 * that is not there is an ordinary answer -- a unit loads the credentials it needs and no
 * others -- while a file that *is* there and cannot be read is a configured source that did not
 * deliver, and falling back from it would turn an unreadable secret into an empty one. fopen
 * reports both as NULL, so errno is what tells them apart, read immediately and once.
 */
static sc_status read_whole_file(const char *path, char *out, size_t out_size, int *absent)
{
    FILE *file;
    size_t read;

    *absent = 0;
    errno = 0;
    file = fopen(path, "rb");
    if (file == NULL) {
        *absent = (errno == ENOENT);
        return SC_ERR_UNAVAILABLE;
    }
    read = fread(out, 1, out_size, file);
    if (read == out_size && fgetc(file) != EOF) {
        fclose(file);
        return SC_ERR_TOO_LONG;
    }
    fclose(file);

    if (read != 0 && out[read - 1] == '\n')
        --read;
    if (read != 0 && out[read - 1] == '\r')
        --read;
    out[read] = '\0';
    return SC_OK;
}

sc_secret_source sc_secret_source_of(const char *name)
{
    const char *credentials;
    char file_variable[128];

    if (name == NULL)
        return SC_SECRET_NOWHERE;

    credentials = getenv("CREDENTIALS_DIRECTORY");
    if (credentials != NULL && credentials[0] != '\0') {
        char path[SC_SECRET_PATH_MAX];
        int written = snprintf(path, sizeof(path), "%s/%s", credentials, name);
        if (written > 0 && (size_t)written < sizeof(path)) {
            /* Opened and closed rather than stat'ed: whether this process can read it is the
             * question, and a mode nobody may read is not a source in force. */
            FILE *file = fopen(path, "rb");
            if (file != NULL) {
                fclose(file);
                return SC_SECRET_CREDENTIAL;
            }
        }
    }

    if (snprintf(file_variable, sizeof(file_variable), "%s_FILE", name) > 0) {
        const char *named = getenv(file_variable);
        if (named != NULL && named[0] != '\0')
            return SC_SECRET_FILE;
    }

    return getenv(name) != NULL ? SC_SECRET_ENVIRONMENT : SC_SECRET_NOWHERE;
}

sc_status sc_secret_read(const char *name, char *out, size_t out_size)
{
    const char *credentials;
    const char *from_env;
    char path[SC_SECRET_PATH_MAX];
    char file_variable[128];
    sc_status status;

    if (name == NULL || out == NULL || out_size == 0)
        return SC_ERR_INVALID_ARGUMENT;
    out[0] = '\0';

    /* 1. systemd. Unset everywhere that is not a systemd service, which is the platform check. */
    credentials = getenv("CREDENTIALS_DIRECTORY");
    if (credentials != NULL && credentials[0] != '\0') {
        int written = snprintf(path, sizeof(path), "%s/%s", credentials, name);
        if (written > 0 && (size_t)written < sizeof(path)) {
            int absent = 0;
            status = read_whole_file(path, out, out_size, &absent);
            if (status == SC_OK)
                return SC_OK;
            /* Absent is the ordinary case -- a unit loads the credentials it needs and no
             * others. Present and unreadable is not, and is the caller's problem to hear
             * about rather than to be quietly given something else. */
            if (!absent) {
                sc_log_fatal(SC_CAT_STARTUP, "config.secret_unreadable",
                             "the systemd credential %s could not be read", path);
                return status == SC_ERR_TOO_LONG ? SC_ERR_TOO_LONG : SC_ERR_UNAVAILABLE;
            }
        }
    }

    /* 2. A file the environment names. */
    if (snprintf(file_variable, sizeof(file_variable), "%s_FILE", name) > 0) {
        const char *named = getenv(file_variable);
        if (named != NULL && named[0] != '\0') {
            int absent = 0;
            /* No exemption for an absent one here: this path was named by a variable somebody
             * set, so "there is no such file" is a broken configuration and not a default. */
            status = read_whole_file(named, out, out_size, &absent);
            if (status == SC_OK)
                return SC_OK;
            sc_log_fatal(SC_CAT_STARTUP, "config.secret_unreadable",
                         "%s names %s, which could not be read -- refusing to fall back to %s",
                         file_variable, named, name);
            return status == SC_ERR_TOO_LONG ? SC_ERR_TOO_LONG : SC_ERR_UNAVAILABLE;
        }
    }

    /* 3. The variable itself. */
    from_env = getenv(name);
    if (from_env == NULL)
        return SC_OK;
    if (strlen(from_env) >= out_size)
        return SC_ERR_TOO_LONG;
    memcpy(out, from_env, strlen(from_env) + 1);
    return SC_OK;
}
